File: csharp/SwimmingCompetitionCS/persistence/RegistrationDBRepository.cs
using System;
using System.Collections.Generic;
using System.Data;
using log4net;
using model;

namespace persistence
{
    public class RegistrationDBRepository : RegistrationRepository
    {
        private static readonly ILog log = LogManager.GetLogger(typeof(RegistrationDBRepository));

        IDictionary<String, string> props;

        public RegistrationDBRepository(IDictionary<string, string> props)
        {
            this.props = props;
        }
        
        public IList<Participant> FindParticipantsByChallenge(Challenge challenge)
        {
            log.InfoFormat("Finding all participants at the challenge {0}", challenge);
            IDbConnection connection = DBUtils.getConnection(props);
            List<Participant> participants = new List<Participant>();
            using (var comm = connection.CreateCommand())
            {
                comm.CommandText = "select * from participant as p inner join registration as r on p.id=r.idparticipant where r.idchallenge = @idchallenge";
                var idChallengeParam = comm.CreateParameter();
                idChallengeParam.ParameterName = "@idchallenge";
                idChallengeParam.Value = challenge.Id;
                comm.Parameters.Add(idChallengeParam);
                
                using (var dataR = comm.ExecuteReader())
                {
                    while (dataR.Read())
                    {
                        long id = dataR.GetInt32(0);
                        String name = dataR.GetString(1);
                        int age = dataR.GetInt16(2);
                        Participant participant = new Participant(name, age);
                        participant.Id = id;
                        participants.Add(participant);
                    }
                }
            }
            log.Info(participants);
            return participants;
        }

        public IList<Challenge> FindChallengesByParticipant(Participant participant)
        {
            log.InfoFormat("Finding all challenges of a participant {0}", participant);
            IDbConnection connection = DBUtils.getConnection(props);
            List<Challenge> challenges = new List<Challenge>();
            using (var comm = connection.CreateCommand())
            {
                comm.CommandText = "select * from challenge as c inner join registration as r on c.id=r.idchallenge where r.idparticipant = @idparticipant";
                var idParticipantParam = comm.CreateParameter();
                idParticipantParam.ParameterName = "@idparticipant";
                idParticipantParam.Value = participant.Id;
                comm.Parameters.Add(idParticipantParam);
                
                using (var dataR = comm.ExecuteReader())
                {
                    while (dataR.Read())
                    {
                        long id = dataR.GetInt32(0);
                        Style style = (Style)Enum.Parse(typeof(Style), dataR.GetString(1));
                        Distance distance = (Distance)dataR.GetInt16(2);
                        Challenge challenge = new Challenge(distance, style);
                        challenge.Id = id;
                        challenges.Add(challenge);
                    }
                }
            }
            log.Info(challenges);
            return challenges;
        }
        
        public void AddParticipantToMoreChallenges(Participant participant, List<Challenge> challenges)
        {
            log.InfoFormat("Saving participants {0} to challenges {1}", participant, challenges);
            var connection = DBUtils.getConnection(props);
            using (var comm = connection.CreateCommand())
            {
                comm.CommandText = "insert into registration(idparticipant, idchallenge) values(@idparticipant,@idchallenge)";
                foreach (var challenge in challenges)
                {
                    comm.Parameters.Clear();
                    var paramIdParticipant = comm.CreateParameter();
                    paramIdParticipant.ParameterName = "@idparticipant";
                    paramIdParticipant.Value = participant.Id;
                    comm.Parameters.Add(paramIdParticipant);

                    var paramIdChallenge = comm.CreateParameter();
                    paramIdChallenge.ParameterName = "@idchallenge";
                    paramIdChallenge.Value = challenge.Id;
                    comm.Parameters.Add(paramIdChallenge);

                    comm.ExecuteNonQuery();
                }
            }
        }
        
        public int CountByChallenge(Challenge challenge)
        {
            log.InfoFormat("Count participants by challenge {0}", challenge);
            IDbConnection connection = DBUtils.getConnection(props);
            using (var comm = connection.CreateCommand())
            {
                comm.CommandText = "select count(*) as count from registration where idchallenge = @idchallenge";
                var idChallengeParam = comm.CreateParameter();
                idChallengeParam.ParameterName = "@idchallenge";
                idChallengeParam.Value = challenge.Id;
                comm.Parameters.Add(idChallengeParam);
                
                using (var dataR = comm.ExecuteReader())
                {
                    while (dataR.Read())
                    {
                        int count = dataR.GetInt16(0);
                        log.InfoFormat("Count {0} participants by challenge {1}", count, challenge);
                        return count;
                    }
                }
            }
            log.InfoFormat("Count 0 participants by challenge {0}", challenge);
            return 0;
        }
        
        private Participant FindParticipant(long along)
        {
            log.InfoFormat("Finding participant with id {0}", along);
            IDbConnection connection = DBUtils.getConnection(props);
            using (var comm = connection.CreateCommand())
            {
                comm.CommandText = "select * from participant where id = @id";
                IDbDataParameter paramId = comm.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = along;
                comm.Parameters.Add(paramId);

                using (var dataR = comm.ExecuteReader())
                {
                    if (dataR.Read())
                    {
                        long id = dataR.GetInt32(0);
                        String name = dataR.GetString(1);
                        int age = dataR.GetInt16(2);
                        Participant participant = new Participant(name, age);
                        participant.Id = id;
                        return participant;
                    }
                }
            }
            log.InfoFormat("Finding NO participant with id {0}", along);
            return null;
        }
        
        private Challenge FindChallenge(long along)
        {  
            log.InfoFormat("Finding a challenge with id {0}", along);
            IDbConnection connection = DBUtils.getConnection(props);
            using (var comm = connection.CreateCommand())
            {
                comm.CommandText = "select * from challenge where id = @id";
                IDbDataParameter paramId = comm.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = along;
                comm.Parameters.Add(paramId);

                using (var dataR = comm.ExecuteReader())
                {
                    if (dataR.Read())
                    {
                        long id = dataR.GetInt32(0);
                        Style style = (Style)Enum.Parse(typeof(Style), dataR.GetString(1));
                        Distance distance = (Distance)dataR.GetInt16(2);
                        Challenge challenge = new Challenge(distance, style);
                        challenge.Id = id;
                        return challenge;
                    }
                }
            }
            log.InfoFormat("Finding NO challenge with id {0}", along);
            return null;
        }
        
        public Registration FindOne(long along)
        {
            log.InfoFormat("Finding registration with id {0}", along);
            IDbConnection connection = DBUtils.getConnection(props);
            using (var comm = connection.CreateCommand())
            {
                comm.CommandText = "select * from registration where id = @id";
                IDbDataParameter paramId = comm.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = along;
                comm.Parameters.Add(paramId);

                using (var dataR = comm.ExecuteReader())
                {
                    if (dataR.Read())
                    {
                        long id = dataR.GetInt32(0);
                        Participant participant = this.FindParticipant(dataR.GetInt32(1));
                        Challenge challenge = this.FindChallenge(dataR.GetInt32(2));
                        Registration registration = new Registration(participant, challenge);
                        registration.Id = id;
                        return registration;
                    }
                }
            }
            log.InfoFormat("Finding NO registration with id {0}", along);
            return null;
        }

        public IEnumerable<Registration> FindAll()
        {
            log.Info("Finding all registrations");
            IDbConnection connection = DBUtils.getConnection(props);
            IList<Registration> registrations = new List<Registration>();
            using (var comm = connection.CreateCommand())
            {
                comm.CommandText = "select * from registration";
                using (var dataR = comm.ExecuteReader())
                {
                    while (dataR.Read())
                    {
                        long id = dataR.GetInt32(0);
                        Participant participant = this.FindParticipant(dataR.GetInt32(1));
                        Challenge challenge = this.FindChallenge(dataR.GetInt32(2));
                        Registration registration = new Registration(participant, challenge);
                        registration.Id = id;
                        registrations.Add(registration);
                    }
                }
            }
            log.Info(registrations);
            return registrations;
        }

        public Registration Save(Registration entity)
        {
            log.InfoFormat("Saving registration {0}", entity);
            var connection = DBUtils.getConnection(props);
            using (var comm = connection.CreateCommand())
            {
                comm.CommandText = "insert into registration(idparticipant, idchallenge) values(@idparticipant,@idchallenge) RETURNING id";
                var paramParticipant = comm.CreateParameter();
                paramParticipant.ParameterName = "@idparticipant";
                paramParticipant.Value = entity.Participant.Id;
                comm.Parameters.Add(paramParticipant);
                
                var paramIdChallenge = comm.CreateParameter();
                paramIdChallenge.ParameterName = "@idchallenge";
                paramIdChallenge.Value = entity.Challenge.Id;
                comm.Parameters.Add(paramIdChallenge);

                long generatedId = (long)comm.ExecuteScalar();
                entity.Id = generatedId;
                return entity;
            }
        }

        public void Delete(long id)
        {
            log.InfoFormat("Deleting registration with id {0}", id);
            IDbConnection connection = DBUtils.getConnection(props);
            using (var comm = connection.CreateCommand())
            {
                comm.CommandText = "delete from registration where id = @id";
                var idParam = comm.CreateParameter();
                idParam.ParameterName = "@id";
                idParam.Value = id;
                comm.Parameters.Add(idParam);
                var result = comm.ExecuteNonQuery();
                if(result == 1)
                    log.InfoFormat("Deleted registration with id {0}", id);
            }
        }

        public void Update(Registration entity)
        {
            //nu e nevoie
        }
    }
}

File: java/SwimmingCompetition/SwimmingCompetitionPersistence/src/main/java/swimmingcompetition/persistence/repository/RegistrationDBRepository.java
package swimmingcompetition.persistence.repository;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import swimmingcompetition.model.*;
import swimmingcompetition.persistence.RegistrationRepository;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class RegistrationDBRepository implements RegistrationRepository {
    private JdbcUtils dbUtils;
    private static final Logger logger= LogManager.getLogger(OrganizerDBRepository.class);
    public RegistrationDBRepository(Properties properties) {
        this.dbUtils = new JdbcUtils(properties);
    }

    @Override
    public List<Challenge> findChallengesByParticipant(Participant participant) {
        logger.traceEntry("finding challenges by participant", participant);
        Connection connection = dbUtils.getConnection();
        List<Challenge> challengeList = new ArrayList<>();
        try(PreparedStatement preparedStatement = connection.prepareStatement("select * from challenge as c inner join registration as r on c.id=r.idchallenge where r.idparticipant = ?")) {
            preparedStatement.setLong(1, participant.getId());
            try(ResultSet resultSet = preparedStatement.executeQuery()) {
                while (resultSet.next()) {
                    Long id = resultSet.getLong("id");
                    Distance distance = Distance.fromInt(resultSet.getInt("distance"));
                    Style style = Style.valueOf(resultSet.getString("style"));
                    Challenge challenge = new Challenge(distance, style);
                    challenge.setId(id);
                    challengeList.add(challenge);
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("Error la DB -- Registration " + e.getMessage().toString());
        }
        logger.traceExit(challengeList);
        return challengeList;
    }

    @Override
    public void addParticipantToMoreChallenges(Participant participant, List<Challenge> challenges) {
        logger.traceEntry("saving registration of participant {} at the challenges {}", participant, challenges);
        Connection connection = dbUtils.getConnection();
        try(PreparedStatement preparedStatement = connection.prepareStatement("insert into registration(idparticipant, idchallenge) values(?,?)")) {
            for(Challenge challenge: challenges) {
                preparedStatement.setLong(1, participant.getId());
                preparedStatement.setLong(2, challenge.getId());
                int result = preparedStatement.executeUpdate();
                Registration registration =  new Registration(participant, challenge);
                ResultSet resultSet = preparedStatement.getGeneratedKeys();
                if (resultSet.next()) {
                    Long id = resultSet.getLong(1);
                    registration.setId(id);
                }
                logger.error("saved registration {}", registration);
            }
        }
        catch (SQLException e) {
            logger.error(e);
            System.out.println("Error la DB -- Registration " + e.getMessage().toString());
        }
    }

    @Override
    public List<Participant> findParticipantsByChallenge(Challenge challenge) {
        logger.traceEntry("finding parcipants by challenge", challenge);
        Connection connection = dbUtils.getConnection();
        List<Participant> participantList = new ArrayList<>();
        try(PreparedStatement preparedStatement = connection.prepareStatement("select * from participant as p inner join registration as r on p.id=r.idparticipant where r.idchallenge = ?")) {
            preparedStatement.setLong(1, challenge.getId());
            try(ResultSet resultSet = preparedStatement.executeQuery()) {
                while (resultSet.next()) {
                    Long id = resultSet.getLong("id");
                    String name = resultSet.getString("name");
                    int age = resultSet.getInt("age");
                    Participant participant = new Participant(name, age);
                    participant.setId(id);
                    participantList.add(participant);
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("Error la DB -- Registration " + e.getMessage());
        }
        logger.traceExit(participantList);
        return participantList;
    }

    @Override
    public int countByChallenge(Challenge challenge) {
        logger.traceEntry("count the participants at the challenge {}",challenge);
        Connection connection = dbUtils.getConnection();
        try(PreparedStatement preparedStatement = connection.prepareStatement("select count(*) as count from registration where idchallenge = ?")) {
            preparedStatement.setLong(1, challenge.getId());
            try(ResultSet resultSet = preparedStatement.executeQuery()) {
                if(resultSet.next()) {
                    int count = resultSet.getInt("count");
                    return count;
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("Error la DB -- Registration " + e.getMessage().toString());
        }
        logger.traceExit("count 0 participants at the challenge {}", challenge);
        return 0;
    }
    private Participant findParticipant(Long aLong) {
        logger.traceEntry("finding participant with id {}", aLong);
        Connection connection = this.dbUtils.getConnection();
        try(PreparedStatement preparedStatement = connection.prepareStatement("select * from participant where id = ?")){
            preparedStatement.setLong(1, aLong);
            try(ResultSet resultSet = preparedStatement.executeQuery()){
                if(resultSet.next()){
                    Long id = resultSet.getLong("id");
                    String name = resultSet.getString("name");
                    int age = resultSet.getInt("age");
                    Participant participant = new Participant(name, age);
                    participant.setId(id);
                    return participant;
                }
            }
        }
        catch (SQLException ex){
            logger.error(ex);
            System.out.println("Eroare la DB -- Registration -- Participant " + ex.getMessage().toString());
        }
        return null;
    }
    private Challenge findChallenge(Long aLong) {
        logger.traceEntry("finding challenge with id {}", aLong);
        Connection connection = this.dbUtils.getConnection();
        try(PreparedStatement preparedStatement = connection.prepareStatement("select * from challenge where id = ?")){
            preparedStatement.setLong(1, aLong);
            try(ResultSet resultSet = preparedStatement.executeQuery()){
                if(resultSet.next()){
                    Long id = resultSet.getLong("id");
                    Style style = Style.valueOf(resultSet.getString("style"));
                    Distance distance = Distance.fromInt(resultSet.getInt("distance"));
                    Challenge challenge = new Challenge(distance, style);
                    challenge.setId(id);
                    return challenge;
                }
            }
        }
        catch (SQLException ex){
            logger.error(ex);
            System.out.println("Eroare la DB -- Challenge " + ex.getMessage().toString());
        }
        logger.traceExit("No challenge found with id {}", aLong);
        return null;
    }
    @Override
    public Registration findOne(Long aLong) {
        //logger.traceEntry("finding registration with id {}", aLong);
        Connection connection = this.dbUtils.getConnection();
        try(PreparedStatement preparedStatement = connection.prepareStatement("select * from registration where id = ?")){
            preparedStatement.setLong(1, aLong);
            try(ResultSet resultSet = preparedStatement.executeQuery()){
                if(resultSet.next()){
                    Long id = resultSet.getLong("id");
                    Long idParticipant = resultSet.getLong("idparticipant");
                    Long idChallenge = resultSet.getLong("idchallenge");
                    Participant participant = this.findParticipant(idParticipant);
                    Challenge challenge = this.findChallenge(idChallenge);
                    Registration registration = new Registration(participant, challenge);
                    return registration;
                }
            }
        }
        catch (SQLException ex){
            logger.error(ex);
            System.out.println("Eroare la DB -- Registration " + ex.getMessage().toString());
        }
       logger.traceExit("finding no registration with id {}", aLong);
        return null;
    }

    @Override
    public Iterable<Registration> findAll() {
        logger.traceEntry("finding all registrations");
        Connection connection = dbUtils.getConnection();
        List<Registration> registrationList = new ArrayList<>();
        try(PreparedStatement preparedStatement = connection.prepareStatement("select * from Registration")) {
            try(ResultSet resultSet = preparedStatement.executeQuery()) {
                while (resultSet.next()) {
                    Long id = resultSet.getLong("id");
                    Long idParticipant = resultSet.getLong("idparticipant");
                    Long idChallenge = resultSet.getLong("idchallenge");
                    Participant participant = this.findParticipant(idParticipant);
                    Challenge challenge = this.findChallenge(idChallenge);
                    Registration registration = new Registration(participant, challenge);
                    registration.setId(id);
                    registrationList.add(registration);
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("Error la DB -- Registration " + e.getMessage().toString());
        }
        logger.traceExit(registrationList);
        return registrationList;
    }

    @Override
    public Registration save(Registration entity) {
        logger.traceEntry("saving registration {}", entity);
        Connection connection = dbUtils.getConnection();
        try(PreparedStatement preparedStatement = connection.prepareStatement("insert into registration(idparticipant, idchallenge) values(?,?)")) {
            preparedStatement.setLong(1, entity.getParticipant().getId());
            preparedStatement.setLong(2, entity.getChallenge().getId());
            int result = preparedStatement.executeUpdate();
            ResultSet resultSet = preparedStatement.getGeneratedKeys();
            if(resultSet.next()){
                Long id = resultSet.getLong(1);
                entity.setId(id);
            }
            logger.error("saved registration {}", entity);
            return entity;
        }
        catch (SQLException e) {
            logger.error(e);
            System.out.println("Error la DB -- Registration " + e.getMessage().toString());
        }
        return null;
    }

    @Override
    public void delete(Long aLong) {
        logger.traceEntry("deleting registration with id {}", aLong);
        Connection connection = dbUtils.getConnection();
        try(PreparedStatement preparedStatement = connection.prepareStatement("delete from registration where id = ?")) {
            preparedStatement.setLong(1, aLong);
            int result = preparedStatement.executeUpdate();
            logger.error("deleted registration {}", result);
        }
        catch (SQLException e) {
            logger.error(e);
            System.out.println("Error la DB -- Registration " + e.getMessage().toString());
        }
    }

    @Override
    public void update(Registration entity) {

    }
}

